compiler/src/lnntoamm/Expr.ts

```diff
@@ -189,6 +189,75 @@ export default abstract class Expr {
       if (prec === -1 || idxs.length === 0) {
         break;
       }
+
+      // all of the selected operators should be the same infix/prefix mode
+      // if the result is null, that means they're not - idk if that's
+      // ever a case so just TODO it
+      const prefixModeOf = (vals: Operator[]) => vals.reduce(
+        (mode, op) => {
+          if (mode === null) return mode;
+          return mode === op.isPrefix ? mode : null;
+        },
+        vals[0].isPrefix
+      );
+
+      idxs.forEach(idx => {
+        const val = precedences[idx];
+        // heat-death-of-the-universe check
+        if (val instanceof Expr) {
+          throw new Error(`uh, how?`)
+        }
+        // ensure that none of the operators disagree on fixity
+        const mode = prefixModeOf(val.get(prec));
+        if (mode === null) {
+          TODO('operator is both prefix and infix - how to determine?');
+        }
+      });
+      // first, prefix operators - we need to mutate idxs so no `.forEach`
```

Comment:
So precedence should still be the first order selection mechanism and then the `expr infix prefix expr` fix is that `prefix` wins out on ties, I think?

Eg if we had an increment prefix operator, and its predecence was below multiplication (and equal addition), then `6 + ++ 5 * 3` should produce `22`, not `24`. (`6 + ++ 5 * 3 -> 6 + ++ 15 -> 6 + 16 -> 22`)

Reply:
(if prefixes are always first regardless of operator predecence, we'd get `6 + ++ 5 * 3 -> 6 + 6 * 3 -> 6 + 18 -> 24`)

Reply:
correct - we don't recalculate `prec` after completing the prefix operator handling so we're still at the same precedence level when we reach the infix operator loop. I just handle all of the prefix operators first to avoid having to repeatedly track forwards and backwards in the same loop